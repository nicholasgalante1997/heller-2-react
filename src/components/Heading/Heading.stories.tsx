import React from 'react';
import { type Meta, type StoryObj } from '@storybook/react';
import { Heading } from './Heading';

const meta: Meta<typeof Heading> = {
  title: 'REACT/Components/Typography/Heading',
  component: Heading
};

export default meta;

type HeadingStory = StoryObj<typeof Heading>;

export const H1: HeadingStory = {
  render: (args) => (
    <Heading {...args} as="h1">
      Heading 100
    </Heading>
  )
};

export const H2: HeadingStory = {
  render: (args) => (
    <Heading {...args} as="h2">
      Heading 200
    </Heading>
  )
};

export const H3: HeadingStory = {
  render: (args) => (
    <Heading {...args} as="h3">
      Heading 300
    </Heading>
  )
};

export const H4: HeadingStory = {
  render: (args) => (
    <Heading {...args} as="h4">
      Heading 400
    </Heading>
  )
};

export const H5: HeadingStory = {
  render: (args) => (
    <Heading {...args} as="h5">
      Heading 500
    </Heading>
  )
};

export const H6: HeadingStory = {
  render: (args) => (
    <Heading {...args} as="h6">
      Heading 600
    </Heading>
  )
};
