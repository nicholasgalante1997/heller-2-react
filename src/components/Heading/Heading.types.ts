import type React from 'react';
interface InternalHeadingProps {
  as: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
}
export type HeadingProps = React.HTMLProps<HTMLHeadingElement> & InternalHeadingProps;
