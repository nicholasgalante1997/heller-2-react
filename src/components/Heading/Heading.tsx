import React from 'react';
import { type HeadingProps } from './Heading.types';
import classNames from 'classnames';

const headingJsxMap = new Map<HeadingProps['as'], React.FC<Omit<HeadingProps, 'as'>>>([
  ['h1', ({ children, ...rest }) => <h1 {...rest}>{children}</h1>],
  ['h2', ({ children, ...rest }) => <h2 {...rest}>{children}</h2>],
  ['h3', ({ children, ...rest }) => <h3 {...rest}>{children}</h3>],
  ['h4', ({ children, ...rest }) => <h4 {...rest}>{children}</h4>],
  ['h5', ({ children, ...rest }) => <h5 {...rest}>{children}</h5>],
  ['h6', ({ children, ...rest }) => <h6 {...rest}>{children}</h6>]
]);

function convertTagToScale(tag: HeadingProps['as']) {
  return tag.replace('h', '');
}

function HeadingComponent({ as, className: classNameProp = '', children, ...rest }: HeadingProps) {
  const Component = React.useMemo(() => headingJsxMap.get(as), [as]);
  const headingClassName = React.useMemo(() => `h-heading-${convertTagToScale(as)}00`, [as]);
  const className = classNames(headingClassName, classNameProp);

  if (!Component) {
    console.error('HELLER-2-REACT:::TYPOGRAPHY:::UNSUPPLIED_AS_PROP');
    return false;
  }

  return (
    <Component className={className} {...rest}>
      {children}
    </Component>
  );
}

export const Heading = React.memo(HeadingComponent as React.FC<HeadingProps>);
Heading.displayName = 'HlrHeading';
