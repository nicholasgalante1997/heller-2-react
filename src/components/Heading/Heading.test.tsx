import React from 'react';
import { render } from '@testing-library/react';
import { Heading } from './Heading';

describe('Heading', () => {
  test('Snapshot', () => {
    const { container } = render(<Heading as="h1">Shmallas</Heading>);
    expect(container).toMatchSnapshot();
  });
});
