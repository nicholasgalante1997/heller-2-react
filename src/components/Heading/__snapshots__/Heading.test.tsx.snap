// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Snapshot 1`] = `
<div>
  <h1
    class="h-heading-100"
  >
    Shmallas
  </h1>
</div>
`;
